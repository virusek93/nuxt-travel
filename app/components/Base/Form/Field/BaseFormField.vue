<script lang="ts" setup>
defineProps<{
  name: string
  label: string
  required?: boolean
  errorMessage?: string
  customClass?: string
}>()
</script>

<template>
  <div>
    <label
      :for="name"
      data-test="form-label"
      class="flex cursor-pointer flex-col items-start gap-0.5 text-sm font-medium text-gray-500"
      :class="customClass"
    >
      <span
        class="after:ml-0.5 after:hidden after:text-red-500 after:content-['*']"
        :class="{ 'after:!inline-block': required }"
      >{{ label }}</span>
    </label>
    <slot />
    <BaseFormErrorMessage
      v-if="errorMessage"
      data-test="error"
    >
      {{ errorMessage }}
    </BaseFormErrorMessage>
  </div>
</template>
