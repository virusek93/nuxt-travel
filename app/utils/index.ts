export const getDateString = (item: Date) => {
  return item.toISOString().split('T')[0]
}
