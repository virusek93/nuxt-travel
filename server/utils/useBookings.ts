import type { Travel } from '~~/types/travel'
import type { Booking } from '~~/types/booking'

export const useBookings = () => {
  const stateKey = 'bookings'
  const idKey = 'bookings-id'

  const getLatestId = async () => {
    return await useStorage().getItem<number>(idKey)
  }

  const addItem = async (item: Omit<Booking, 'id'>) => {
    let latestId = await getLatestId()
    if (!latestId) {
      latestId = 0
    }
    const updatedLatestId = latestId + 1
    await useStorage(stateKey).setItem(updatedLatestId.toString(), {
      ...item,
      id: updatedLatestId,
    })
    await useStorage().setItem(idKey, updatedLatestId)
    return true
  }

  const getBooking = async (id: string) => {
    return await useStorage(stateKey).getItem(id)
  }

  const getAllBookings = async () => {
    const keys = await useStorage(stateKey).getKeys()
    const prefixedKeys = keys.map(item => `${stateKey}:${item}`)
    return await useStorage().getItems<Travel>(prefixedKeys)
  }

  const deleteItem = async (id: string) => {
    if (!(await useStorage(stateKey).hasItem(id))) {
      return false
    }
    await useStorage(stateKey).removeItem(id)
    return true
  }

  const updateItem = async (id: string, item: Travel) => {
    await useStorage(stateKey).setItem(id.toString(), { ...item, id })
    return true
  }

  return { addItem, getBooking, deleteItem, updateItem, getAllBookings }
}
