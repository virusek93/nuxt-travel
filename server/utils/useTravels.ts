import type { Travel } from '~~/types/travel'

export const useTravels = () => {
  const stateKey = 'user-travel'
  const idKey = 'user-travel-id'

  const getLatestId = async () => {
    return await useStorage().getItem<number>(idKey)
  }

  const addItem = async (item: Omit<Travel, 'id'>) => {
    let latestId = await getLatestId()
    if (!latestId) {
      latestId = 0
    }
    const updatedLatestId = latestId + 1
    await useStorage(stateKey).setItem(updatedLatestId.toString(), {
      ...item,
      id: updatedLatestId,
    })
    await useStorage().setItem(idKey, updatedLatestId)
    return true
  }

  const getTravel = async (id: string) => {
    return await useStorage(stateKey).getItem(id)
  }

  const getAllTravels = async () => {
    const keys = await useStorage(stateKey).getKeys()
    const prefixedKeys = keys.map(item => `${stateKey}:${item}`)
    return await useStorage().getItems<Travel>(prefixedKeys)
  }

  const deleteItem = async (id: string) => {
    if (!(await useStorage(stateKey).hasItem(id))) {
      return false
    }
    await useStorage(stateKey).removeItem(id)
    return true
  }

  const updateItem = async (id: string, item: Travel) => {
    await useStorage(stateKey).setItem(id.toString(), { ...item, id })
    return true
  }

  return { addItem, getTravel, deleteItem, updateItem, getAllTravels }
}
