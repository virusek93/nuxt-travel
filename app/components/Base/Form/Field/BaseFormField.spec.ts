import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import BaseFormField from '~/components/Base/Form/Field/BaseFormField.vue'
import { getTestElement } from '~/utils/tests'

const props: InstanceType<typeof BaseFormField>['$props'] = {
  name: 'testName',
  label: 'testLabel',
  errorMessage: 'testErrorMessage',
}

describe('BaseFormField', () => {
  it('Should render and match the snapshot', () => {
    const wrapper = mount(BaseFormField, { props })
    expect(wrapper.element).toMatchSnapshot()
  })
  it('Should set name and render label text', () => {
    const wrapper = mount(BaseFormField, { props })
    const label = wrapper.find(getTestElement('form-label'))
    expect(label.attributes('for')).toBe(props.name)
    expect(label.text()).toBe(props.label)
  })
  it('Should render errorMessage only if errorMessage exists', async () => {
    const wrapper = mount(BaseFormField, { props })
    const errorMessage = wrapper.find(getTestElement('error'))
    expect(errorMessage.exists()).toBeTruthy()
    expect(errorMessage.text()).toBe(props.errorMessage)
    await wrapper.setProps({ errorMessage: undefined })
    expect(wrapper.find(getTestElement('error')).exists()).toBeFalsy()
  })
})
