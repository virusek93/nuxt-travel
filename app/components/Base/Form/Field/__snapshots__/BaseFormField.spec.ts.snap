// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseFormField > Should render and match the snapshot 1`] = `
<div>
  <label
    class="flex cursor-pointer flex-col items-start gap-0.5 text-sm font-medium text-gray-500"
    data-test="form-label"
    for="testName"
  >
    <span
      class="after:ml-0.5 after:hidden after:text-red-500 after:content-['*']"
    >
      testLabel
    </span>
  </label>
  
  
  <span
    class="text-sm text-red-500"
    data-test="error"
  >
    
    testErrorMessage
    
  </span>
</div>
`;
